DSESwift/DataSyncer.swift

```diff
@@ -8,37 +8,398 @@
+                    self.timer = NSTimer.scheduledTimerWithTimeInterval(self.syncRate, target: self, selector: "sync", userInfo: nil, repeats: true);
+                }
+                self.timer!.fire()
+            }else{
+                print("DSE is not initialized")
+            }
+        })
+    }
+
+    func disablePeriodicSync(){
+        if self.timer == nil {
+            timer!.invalidate()
+        } else {
+            print("timer is nil")
+        }
+    }
+    
+    /**
+     * Synchronize data in local and remote storage.
+     * Is executed asynchronously.
+     */
+    public func sync() {
+        
+        dispatch_promise(on: data_syncer_process_queue, body:{
+            return Promise<Void> { fulfill, reject in
+                // step 1
+                try self.processDeletionRequests()
+                // step 2
+                try self.uploadSensorDataToRemote()
+                // step 3
+                try self.downloadSensorsFromRemote()
+                // step 4
+                try self.downloadSensorsDataFromRemote()
+                // step 5
+                try self.cleanLocalStorage()
+                fulfill()
+            }
+        }).then({
+            self.delegate?.onSyncCompleted()
+        }).error({ error in
+            print(error)
+            self.delegate?.onSyncFailed(error)
+        })
+    }
+    
+    func downloadSensorProfiles() throws {
+        let sensorProfiles = try SensorDataProxy.getSensorProfiles()
+        for ( _ ,subJson):(String, JSON) in sensorProfiles {
+            let (sensorName, structure) = self.getSensorNameAndStructure(subJson)
+            try DatabaseHandler.createOrUpdateSensorProfile(sensorName, dataStructure: structure)
+        }
+    }
+
+    
+    func processDeletionRequests() throws {
+        let dataDeletionRequests = DatabaseHandler.getDataDeletionRequests()
+        for request in dataDeletionRequests {
+            try SensorDataProxy.deleteSensorData(sourceName: request.sourceName, sensorName: request.sensorName, startTime: request.startTime, endTime: request.endTime)
+            // remove the deletion request which has been processed
+            try DatabaseHandler.deleteDataDeletionRequest(request.uuid)
+        }
+    }
+    
+    
+    func downloadSensorsFromRemote() throws {
+        do{
+            let sensors = try getSensorsFromRemote()
+            try insertSensorsIntoLocal(sensors)
+            //notify
+            self.delegate?.onSensorsDownloadCompleted()
+        }catch {
+            self.delegate?.onSensorDataDownloadFailed(error)
+            throw error
+        }
+
+    }
+    
+    func downloadSensorsDataFromRemote() throws {
+        do{
+            let sensorsInLocal = getSensorsInLocal()
+            for sensor in sensorsInLocal{
+                if (sensor.remoteDownloadEnabled){
+                    try downloadAndStoreDataForSensor(sensor)
+                    try updateDownloadStatusForSensor(sensor)
```

Comment:
the second method is very small and only used here. Put it in here with an explanatory comment to reduce code. methods should be used for readability and or avoiding code reuse. Both of these cases are not really relevant for the updateDownloadStatusForSensor method.